### DISCUSSION TITLE
Timer is not running accurately

### DISCUSSION THREAD

**Discussion by oldnick85** [libuv/libuv#4639] [Q&A] [answered]
Hi all!

I use a repeating timer with a fixed timeout, but the intervals between the timer triggers are slightly longer than I expect. This code [https://github.com/oldnick85/check_libuv_timer](https://github.com/oldnick85/check_libuv_timer) gives the following:
> start
1: 0
2: 100
3: 200
...
9: 800
10: 901
...
18: 1701
19: 1802
...
96: 9509
97: 9610
98: 9710
99: 9810
100: 9910
stop

Obviously, additional delay is accumulating. Is it a bug or a feature? how can i get accurate timer triggers without delay accumulation?

Environmemt: Ubuntu 24.04.1 LTS, gcc 13.2.0, libuv1-dev: 1.48.0-1.1build1

**Comment by bnoordhuis** [comment, marked as answer]
That's expected; libuv does not track overhead of callbacks, etc. If that's the behavior you want, start and stop the timer manually and use uv_hrtime() to compute the timeout.

**Reply by oldnick85** [reply]
Now it's clear. I implemented the behavior I needed using C++ (since I use it myself) in [the same repository](https://github.com/oldnick85/check_libuv_timer) (if this helps anyone).

But it seems strange to me that without additional algorithms I cannot implement, for example, sending an RTP stream of exactly 50 packets per second.

Perhaps it is worth adding different modes of timer working or at least explicitly note such nuance of timer behavior in the documentation?

Thanks for answer!

**Reply by bnoordhuis** [reply]
It's already mentioned in the timers documentation but only explicitly for uv_timer_set_repeat. I've opened #4640 to clarify.

> is worth adding different modes of timer working

It's been discussed ad nauseam in the past but the problem is you need to come up with an overrun policy (ex. what do you do when the callback takes longer to complete than the interval the timer ticks at?) and since there's no single good answer, we punt it to the libuv consumer.
